tests/e2e/json_utils.go

```diff
@@ -0,0 +1,412 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+)
+
+// MarshalJSON marshals a step into JSON while including the type of the action.
+func (step Step) MarshalJSON() ([]byte, error) {
+	actionType := reflect.TypeOf(step.Action)
+
+	result := struct {
+		ActionType string
+		Action     interface{}
+		State      State
+	}{
+		ActionType: actionType.String(),
+		Action:     step.Action,
+		State:      step.State,
+	}
+
+	return json.Marshal(result)
+}
+
+// UnmarshalJSON unmarshals a step from JSON while including the type of the action.
+func (step *Step) UnmarshalJSON(data []byte) error {
+	var tmp struct {
+		ActionType string
+		Action     json.RawMessage
+		State      State
+	}
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+
+	action, err := UnmarshalMapToActionType(tmp.Action, tmp.ActionType)
```

Comment:
Just to note but it seems peculiar to me that we need a specific `UnmarshalMapToActionType` but when marshalling in `MarshalJSON()` we do not need any `marshalAction` function.

Reply:
It is implicit: See how in the steps, I am modifying the definition of step to include a different action.
It might be doable to put the actionType inside the action, too, and modify the way actions are unmarshalled and marshalled, would you prefer that?